### ISSUE TITLE
Convert stdlib dataclass to a pydantic BaseModel

### ISSUE THREAD

**Issue by samiashi** [pydantic/pydantic#5158]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

I'm trying to convert a stdlib dataclass to Pydantic BaseModel dynamically and I've seen this answer https://github.com/pydantic/pydantic/issues/1967#issuecomment-742698281

But when I use this approach I have `mypy` complain when I use this class as `def something() -> A:`
```
error: Variable "A" is not valid as a type
note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
```

Any ideas on how to resolve this? I wanna be able to use A as an object & type if possible

### Example Code

```Python
from typing import Type

from pydantic import BaseModel
from pydantic.dataclasses import dataclass as pydantic_dataclass

from somelib.models import A as SomelibA
# @dataclass
# class A:
#   foo: str
#   bar: str


def model_from_dataclass(kls: Type) -> Type[BaseModel]:
    """Converts a stdlib dataclass to a pydantic BaseModel"""
    return pydantic_dataclass(kls).__pydantic_model__


A = model_from_dataclass(SomelibA)

# Ideally I want A to have the same functionality as
# class A(BaseModel):
#   foor: str
#   bar: str
```

### Python, Pydantic & OS Version

```Text
python 3.9.14
pydantic 1.10.2
mypy 0.961
```

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/usage/schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/usage/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/usage/model_config/)
- [ ] [Field Types](https://docs.pydantic.dev/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/usage/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by dmontagu** [comment]
I don't think this is going to be compatible with mypy, and will be very challenging to make work as you intend. I wouldn't consider this a bug in pydantic considering this is expected behavior from mypy.

The closest we could do would probably be to add something like `model_from_dataclass` to the library as a utility function, and augment the mypy plugin to do what you are expecting.

I am not sure how difficult it would be to augment the plugin to be aware of this; I could _imagine_ it being fairly straightforward, in which case I would be open to accepting a PR making it work as long as it looked like it wouldn't be a significant maintenance burden.

More realistically, I expect this would be fairly complex, in which case I think it probably wouldn't be worth the maintenance burden unless there was very broad interest.

**Comment by samuelcolvin** [comment]
Agreed
